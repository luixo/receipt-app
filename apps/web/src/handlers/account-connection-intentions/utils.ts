import { TRPCError } from "@trpc/server";
import type { z } from "zod";

import { ACCOUNT_CONNECTIONS_INTENTIONS } from "~db/consts";
import type { AccountsId, UsersId } from "~db/models";
import type { Database } from "~db/types";
import type { emailSchema } from "~web/handlers/validation";

export const addConnectionIntention = async (
	database: Database,
	fromAccountId: AccountsId,
	user: { name: string },
	toEmail: z.infer<typeof emailSchema>,
	asUserId: UsersId,
) => {
	const targetAccount = await database
		.selectFrom("accounts")
		.select(["id", "avatarUrl"])
		.where("email", "=", toEmail.lowercase)
		.executeTakeFirst();
	if (!targetAccount) {
		throw new TRPCError({
			code: "NOT_FOUND",
			message: `Account with email "${toEmail.original}" does not exist.`,
		});
	}
	const connectedUser = await database
		.selectFrom("accounts")
		.where("accounts.email", "=", toEmail.lowercase)
		.innerJoin("users", (qb) =>
			qb
				.onRef("users.connectedAccountId", "=", "accounts.id")
				.on("users.ownerAccountId", "=", fromAccountId),
		)
		.select("users.name")
		.executeTakeFirst();
	if (connectedUser) {
		throw new TRPCError({
			code: "CONFLICT",
			message: `Account with email "${toEmail.original}" is already connected to user "${connectedUser.name}".`,
		});
	}
	const viceVersaIntention = await database
		.selectFrom("accountConnectionsIntentions")
		.select("userId")
		.where((eb) =>
			eb.and({
				accountId: targetAccount.id,
				targetAccountId: fromAccountId,
			}),
		)
		.executeTakeFirst();
	if (viceVersaIntention) {
		await database.transaction().execute(async (tx) => {
			await tx
				.updateTable("users")
				.set({ connectedAccountId: fromAccountId })
				.where((eb) =>
					eb.and({
						ownerAccountId: targetAccount.id,
						id: viceVersaIntention.userId,
					}),
				)
				.executeTakeFirst();
			await tx
				.updateTable("users")
				.set({ connectedAccountId: targetAccount.id })
				.where((eb) =>
					eb.and({
						ownerAccountId: fromAccountId,
						id: asUserId,
					}),
				)
				.executeTakeFirst();
			await tx
				.deleteFrom("accountConnectionsIntentions")
				.where((eb) =>
					eb.and({
						accountId: targetAccount.id,
						targetAccountId: fromAccountId,
					}),
				)
				.executeTakeFirst();
		});
		return {
			account: {
				id: targetAccount.id,
				email: toEmail.lowercase,
				avatarUrl: targetAccount.avatarUrl || undefined,
			},
			connected: true,
			user: {
				name: user.name,
			},
		};
	}
	try {
		await database
			.insertInto("accountConnectionsIntentions")
			.values({
				accountId: fromAccountId,
				targetAccountId: targetAccount.id,
				userId: asUserId,
				createdAt: new Date(),
			})
			.executeTakeFirst();
		return {
			account: {
				id: targetAccount.id,
				email: toEmail.lowercase,
				avatarUrl: targetAccount.avatarUrl || undefined,
			},
			connected: false,
			user: {
				name: user.name,
			},
		};
	} catch (e) {
		// Could be like `duplicate key value violates unique constraint "..."`
		const message = String(e);
		if (
			message.includes(ACCOUNT_CONNECTIONS_INTENTIONS.CONSTRAINTS.ACCOUNT_PAIR)
		) {
			const existingIntention = await database
				.selectFrom("accountConnectionsIntentions")
				.where((eb) =>
					eb.and({
						accountId: fromAccountId,
						targetAccountId: targetAccount.id,
					}),
				)
				.select(["userId"])
				.executeTakeFirstOrThrow();
			const existingUser = await database
				.selectFrom("users")
				.select("name")
				.where("id", "=", existingIntention.userId)
				.executeTakeFirstOrThrow();
			throw new TRPCError({
				code: "CONFLICT",
				message: `You already has intention to connect to "${toEmail.original}" as user "${existingUser.name}".`,
			});
		}
		if (
			message.includes(ACCOUNT_CONNECTIONS_INTENTIONS.CONSTRAINTS.USER_PAIR)
		) {
			throw new TRPCError({
				code: "CONFLICT",
				message: `You already has intention to connect to user "${user.name}".`,
			});
			// This is probably a c8 bug
			/* c8 ignore next */
		}
		/* c8 ignore next 2 */
		throw e;
	}
};
