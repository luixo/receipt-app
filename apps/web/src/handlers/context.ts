import { type H3Event, createEvent } from "@tanstack/react-start/server";
import type { inferProcedureBuilderResolverOptions } from "@trpc/server";
import type { TRPCRequestInfo } from "@trpc/server/http";
import type { IncomingMessage, ServerResponse } from "node:http";

import type { Database } from "~db/database";
import type { TestContext } from "~tests/backend/utils/test";
import type { authProcedure } from "~web/handlers/trpc";
import type { CacheDbOptions } from "~web/providers/cache-db";
import type { EmailOptions } from "~web/providers/email";
import type { ExchangeRateOptions } from "~web/providers/exchange-rate";
import type { Logger } from "~web/providers/logger";
import type { S3Options } from "~web/providers/s3";

type TestContextPicks = Pick<TestContext, "getSalt" | "getUuid"> & {
	database: Database;
	logger: Logger;
	emailOptions: EmailOptions;
	cacheDbOptions: CacheDbOptions;
	exchangeRateOptions: ExchangeRateOptions;
	s3Options: S3Options;
};

export type NetContext = {
	event: H3Event;
	info: TRPCRequestInfo;
};

export type UnauthorizedContext = NetContext & TestContextPicks;

export type AuthorizedContext = inferProcedureBuilderResolverOptions<
	typeof authProcedure
>["ctx"];

export const createContext = (
	{
		req,
		res,
		info,
	}: { req: IncomingMessage; res: ServerResponse; info: TRPCRequestInfo },
	testContext: TestContextPicks,
): UnauthorizedContext => ({
	info,
	event: createEvent(req, res),
	...testContext,
});
