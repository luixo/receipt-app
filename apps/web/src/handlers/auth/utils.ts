import { TRPCError } from "@trpc/server";

import { DAY } from "~utils";
import type { AccountsId, SessionsSessionId } from "~web/db/models";
import { generateConfirmEmailEmail } from "~web/email/utils";
import type { UnauthorizedContext } from "~web/handlers/context";
import { getEmailClient } from "~web/providers/email";

// How long a session should last
export const SESSION_EXPIRATION_DURATION = 30 * DAY;
// How long session should be intact before we update it
export const SESSION_SHOULD_UPDATE_EVERY = 2 * DAY;

export const getExpirationDate = () =>
	new Date(Date.now() + SESSION_EXPIRATION_DURATION);

export const createAuthorizationSession = async (
	ctx: UnauthorizedContext,
	accountId: AccountsId,
) => {
	const uuid: SessionsSessionId = ctx.getUuid();
	const expirationDate = getExpirationDate();
	await ctx.database
		.insertInto("sessions")
		.values({
			accountId,
			sessionId: uuid,
			expirationTimestamp: expirationDate,
		})
		.executeTakeFirst();
	return {
		authToken: uuid,
		expirationDate,
	};
};

export const sendVerificationEmail = async (
	ctx: UnauthorizedContext,
	email: string,
	token: string,
) => {
	if (!ctx.emailOptions.active) {
		throw new TRPCError({
			code: "FORBIDDEN",
			message: "Currently email resend is not supported",
		});
	}
	try {
		await getEmailClient(ctx).send({
			address: email,
			subject: "Confirm email in Receipt App",
			body: generateConfirmEmailEmail(ctx, token),
		});
	} catch (e) {
		throw new TRPCError({
			code: "INTERNAL_SERVER_ERROR",
			message: `Something went wrong: ${
				/* c8 ignore next */
				e instanceof Error ? e.message : String(e)
			}`,
		});
	}
};
