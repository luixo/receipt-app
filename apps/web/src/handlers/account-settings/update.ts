import { z } from "zod";

import type { SimpleUpdateObject } from "~db/types";
import { authProcedure } from "~web/handlers/trpc";

import { DEFAULT_ACCOUNT_SETTINGS } from "./get";

type SettingsUpdateObject = SimpleUpdateObject<"accountSettings">;

export const procedure = authProcedure
	.input(
		z.discriminatedUnion("type", [
			z.strictObject({
				type: z.literal("manualAcceptDebts"),
				value: z.boolean(),
			}),
		]),
	)
	.mutation(async ({ ctx, input }) => {
		const { database } = ctx;
		const updateObject: SettingsUpdateObject = {};
		switch (input.type) {
			// We want this to blow up in case we add more cases
			// eslint-disable-next-line @typescript-eslint/no-unnecessary-condition
			case "manualAcceptDebts":
				updateObject.manualAcceptDebts = input.value;
		}
		const existingSettings = await database
			.selectFrom("accountSettings")
			.where("accountSettings.accountId", "=", ctx.auth.accountId)
			.limit(1)
			.executeTakeFirst();
		if (!existingSettings) {
			await database
				.insertInto("accountSettings")
				.values({
					accountId: ctx.auth.accountId,
					...DEFAULT_ACCOUNT_SETTINGS,
					...updateObject,
				})
				.executeTakeFirst();
		} else {
			await database
				.updateTable("accountSettings")
				.set(updateObject)
				.where("accountSettings.accountId", "=", ctx.auth.accountId)
				.executeTakeFirst();
		}
	});
