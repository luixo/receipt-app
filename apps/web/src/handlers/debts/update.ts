import { TRPCError } from "@trpc/server";
import { isNonNullish, keys, omitBy } from "remeda";
import { z } from "zod";

import { debtAmountSchema, debtNoteSchema } from "~app/utils/validation";
import type { SimpleUpdateObject } from "~db/types";
import { queueCallFactory } from "~web/handlers/batch";
import type { AuthorizedContext } from "~web/handlers/context";
import { authProcedure } from "~web/handlers/trpc";
import {
	currencyCodeSchema,
	debtIdSchema,
	receiptIdSchema,
} from "~web/handlers/validation";

import { upsertAutoAcceptedDebts } from "./utils";

const updateDebtSchema = z.strictObject({
	id: debtIdSchema,
	update: z
		.strictObject({
			amount: debtAmountSchema,
			timestamp: z.date(),
			note: debtNoteSchema,
			currencyCode: currencyCodeSchema,
			locked: z.boolean(),
			receiptId: receiptIdSchema.optional(),
		})
		.partial()
		.refine(
			(obj) => keys(obj).length !== 0,
			"Update object has to have at least one key to update",
		),
});

type DebtUpdateObject = SimpleUpdateObject<"debts">;

const KEYS_NOT_UPDATE_LOCKED_TIMESTAMP: (keyof DebtUpdateObject)[] = [
	"note",
	"receiptId",
];

const buildSetObjects = (
	input: z.infer<typeof updateDebtSchema>,
	debt: { lockedTimestamp: Date | null },
) => {
	const setObject: DebtUpdateObject = omitBy(
		{
			amount: input.update.amount?.toString(),
			timestamp: input.update.timestamp,
			note: input.update.note,
			currencyCode: input.update.currencyCode,
			receiptId: input.update.receiptId,
			lockedTimestamp:
				input.update.locked === undefined
					? debt.lockedTimestamp === null
						? undefined
						: new Date()
					: input.update.locked
					? new Date()
					: null,
		},
		(value) => value === undefined,
	);
	const keysToUpdateLockedTimestamp = keys(input.update).filter(
		(key) =>
			!KEYS_NOT_UPDATE_LOCKED_TIMESTAMP.includes(key as keyof DebtUpdateObject),
	);
	if (keysToUpdateLockedTimestamp.length === 0) {
		// don't update lockedTimestamp if we updated keys that don't require lockedTimestamp update
		delete setObject.lockedTimestamp;
	}
	const reverseSetObject = omitBy(
		{
			...setObject,
			note: undefined,
			amount:
				input.update.amount === undefined
					? undefined
					: (-input.update.amount).toString(),
		},
		(value) => value === undefined,
	);
	return {
		setObject,
		reverseSetObject:
			keys(reverseSetObject).length === 0 ? undefined : reverseSetObject,
	};
};

const fetchDebts = async (
	ctx: AuthorizedContext,
	updates: readonly z.infer<typeof updateDebtSchema>[],
) =>
	ctx.database
		.selectFrom("debts")
		.where((eb) =>
			eb(
				"debts.id",
				"in",
				updates.map((update) => update.id),
			).and("debts.ownerAccountId", "=", ctx.auth.accountId),
		)
		.innerJoin("users", (qb) =>
			qb
				.onRef("users.id", "=", "debts.userId")
				.onRef("users.ownerAccountId", "=", "debts.ownerAccountId"),
		)
		.leftJoin("accountSettings", (qb) =>
			qb.onRef("users.connectedAccountId", "=", "accountSettings.accountId"),
		)
		.leftJoin("users as usersTheir", (qb) =>
			qb
				.onRef("usersTheir.connectedAccountId", "=", "debts.ownerAccountId")
				.onRef("usersTheir.ownerAccountId", "=", "users.connectedAccountId"),
		)
		.select([
			"debts.id",
			"debts.userId",
			"debts.lockedTimestamp",
			"debts.note",
			"debts.currencyCode",
			"debts.amount",
			"debts.timestamp",
			"debts.receiptId",
			"users.connectedAccountId as foreignAccountId",
			"accountSettings.manualAcceptDebts",
			"usersTheir.id as theirUserId",
		])
		.execute();

const isError = <T>(input: T | TRPCError): input is TRPCError =>
	input instanceof TRPCError;

type Debt = Awaited<ReturnType<typeof fetchDebts>>[number];

const getUpdatedDebtsWithErrors = (
	updatesWithErrors: (
		| {
				update: z.infer<typeof updateDebtSchema>;
				debt: Debt;
		  }
		| TRPCError
	)[],
) =>
	updatesWithErrors.map((updateOrError) => {
		if (isError(updateOrError)) {
			return updateOrError;
		}
		const { debt, update } = updateOrError;
		return { ...buildSetObjects(update, debt), debt };
	});

const updateAutoAcceptingDebts = async (
	ctx: AuthorizedContext,
	updatedDebtsWithErrors: ReturnType<typeof getUpdatedDebtsWithErrors>,
) => {
	const autoAcceptedData = updatedDebtsWithErrors
		.map((updatedDebtOrError) => {
			if (isError(updatedDebtOrError)) {
				return null;
			}
			const { debt, setObject, reverseSetObject } = updatedDebtOrError;
			if (
				!debt.manualAcceptDebts &&
				debt.foreignAccountId &&
				debt.theirUserId
			) {
				return {
					id: debt.id,
					ownerAccountId: debt.foreignAccountId,
					userId: debt.theirUserId,
					currencyCode: debt.currencyCode,
					amount: (-Number(debt.amount)).toString(),
					timestamp: debt.timestamp,
					lockedTimestamp: debt.lockedTimestamp,
					receiptId: debt.receiptId,
					created: new Date(),
					...reverseSetObject,
					// In case debt doesn't exist - we need to set a new note, not the old one
					note: setObject.note || debt.note,
					isNew: false,
				};
			}
			return null;
		})
		.filter(isNonNullish);
	if (autoAcceptedData.length === 0) {
		return [];
	}
	const { newDebts } = await upsertAutoAcceptedDebts(
		ctx.database,
		autoAcceptedData,
	);
	return updatedDebtsWithErrors
		.map((updatedDebtOrError) => {
			if (
				isError(updatedDebtOrError) ||
				(updatedDebtOrError.reverseSetObject?.lockedTimestamp === undefined &&
					!newDebts.some((debt) => updatedDebtOrError.debt.id === debt.id))
			) {
				return;
			}
			return updatedDebtOrError.debt.id;
		})
		.filter(isNonNullish);
};

const updateDebts = async (
	ctx: AuthorizedContext,
	updatedDebtsWithErrors: ReturnType<typeof getUpdatedDebtsWithErrors>,
) =>
	ctx.database.transaction().execute((tx) =>
		Promise.all(
			updatedDebtsWithErrors.map((updatedDebtOrError) => {
				if (isError(updatedDebtOrError)) {
					return null;
				}
				return tx
					.updateTable("debts")
					.set(updatedDebtOrError.setObject)
					.where((eb) =>
						eb.and({
							id: updatedDebtOrError.debt.id,
							ownerAccountId: ctx.auth.accountId,
						}),
					)
					.executeTakeFirst();
			}),
		),
	);

const queueUpdateDebt = queueCallFactory<
	AuthorizedContext,
	z.infer<typeof updateDebtSchema>,
	{
		// value or null for set object, undefined for not being set
		lockedTimestamp: Date | null | undefined;
		reverseLockedTimestampUpdated: boolean;
	}
>((ctx) => async (updates) => {
	const debts = await fetchDebts(ctx, updates);
	const updatesWithErrors = updates.map((update) => {
		const matchedDebt = debts.find((debt) => debt.id === update.id);
		if (!matchedDebt) {
			return new TRPCError({
				code: "NOT_FOUND",
				message: `Debt "${update.id}" does not exist on account "${ctx.auth.email}".`,
			});
		}
		return { update, debt: matchedDebt };
	});
	const errors = updatesWithErrors.filter(isError);
	if (updatesWithErrors.length === errors.length) {
		return errors;
	}
	const updatedDebtsWithErrors = getUpdatedDebtsWithErrors(updatesWithErrors);
	const updatedDebtIds = await updateAutoAcceptingDebts(
		ctx,
		updatedDebtsWithErrors,
	);
	await updateDebts(ctx, updatedDebtsWithErrors);
	return updatedDebtsWithErrors.map((updatedDebtOrError) => {
		if (isError(updatedDebtOrError)) {
			return updatedDebtOrError;
		}
		return {
			lockedTimestamp: updatedDebtOrError.setObject.lockedTimestamp,
			reverseLockedTimestampUpdated: updatedDebtIds.includes(
				updatedDebtOrError.debt.id,
			),
		};
	});
});

export const procedure = authProcedure
	.input(updateDebtSchema)
	.mutation(queueUpdateDebt);
