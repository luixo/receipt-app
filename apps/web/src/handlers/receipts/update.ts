import { TRPCError } from "@trpc/server";
import { z } from "zod";

import { receiptNameSchema } from "~app/utils/validation";
import type { SimpleUpdateObject } from "~db/types";
import { authProcedure } from "~web/handlers/trpc";
import { currencyCodeSchema, receiptIdSchema } from "~web/handlers/validation";

type ReceiptUpdateObject = SimpleUpdateObject<"receipts">;

export const procedure = authProcedure
	.input(
		z.strictObject({
			id: receiptIdSchema,
			update: z.discriminatedUnion("type", [
				z.strictObject({
					type: z.literal("name"),
					name: receiptNameSchema,
				}),
				z.strictObject({ type: z.literal("issued"), issued: z.date() }),
				z.strictObject({
					type: z.literal("currencyCode"),
					currencyCode: currencyCodeSchema,
				}),
			]),
		}),
	)
	.mutation(async ({ input, ctx }) => {
		const { database } = ctx;
		const receipt = await database
			.selectFrom("receipts")
			.select(["ownerAccountId"])
			.where("id", "=", input.id)
			.executeTakeFirst();
		if (!receipt) {
			throw new TRPCError({
				code: "NOT_FOUND",
				message: `No receipt found by id "${input.id}".`,
			});
		}
		if (receipt.ownerAccountId !== ctx.auth.accountId) {
			throw new TRPCError({
				code: "FORBIDDEN",
				message: `Receipt "${input.id}" is not owned by "${ctx.auth.email}".`,
			});
		}
		const setObject: ReceiptUpdateObject = {};
		switch (input.update.type) {
			case "currencyCode":
				setObject.currencyCode = input.update.currencyCode;
				break;
			case "issued":
				setObject.issued = input.update.issued;
				break;
			case "name":
				setObject.name = input.update.name;
				break;
		}

		await database
			.updateTable("receipts")
			.set(setObject)
			.where("id", "=", input.id)
			.executeTakeFirst();
	});
