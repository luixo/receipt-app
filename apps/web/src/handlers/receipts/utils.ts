import type { QueryCreator } from "kysely";
import type { z } from "zod";

import type { AccountsId } from "~db/models";
import type { Database, ReceiptsDatabase } from "~db/types";
import type { assignableRoleSchema } from "~web/handlers/validation";
import { roleSchema } from "~web/handlers/validation";

export type Role = z.infer<typeof roleSchema>;
export type AssignableRole = z.infer<typeof assignableRoleSchema>;

export const getAccessRole = async (
	database: Database,
	receipt: { ownerAccountId: string; id: string },
	accountId: string,
): Promise<Role | undefined> => {
	if (receipt.ownerAccountId === accountId) {
		return "owner";
	}
	const participant = await database
		.selectFrom("receiptParticipants")
		.innerJoin("users", (jb) =>
			jb.onRef("users.id", "=", "receiptParticipants.userId"),
		)
		.innerJoin("accounts", (jb) =>
			jb.onRef("accounts.id", "=", "users.connectedAccountId"),
		)
		.where((eb) =>
			eb.and({
				"accounts.id": accountId,
				receiptId: receipt.id,
			}),
		)
		.select("receiptParticipants.role")
		.limit(1)
		.executeTakeFirst();
	if (!participant) {
		return;
	}
	const parsed = roleSchema.safeParse(participant.role);
	/* c8 ignore start */
	if (!parsed.success) {
		// TODO: add database-level validation
		return;
	}
	/* c8 ignore stop */
	return parsed.data;
};

export const getOwnReceipts = (
	database: Database | QueryCreator<ReceiptsDatabase>,
	ownerAccountId: AccountsId,
) =>
	database.selectFrom("receipts").where("ownerAccountId", "=", ownerAccountId);

export const getParticipantsReceipts = (
	database: Database | QueryCreator<ReceiptsDatabase>,
	ownerAccountId: AccountsId,
) =>
	database
		.selectFrom("users")
		.where((eb) =>
			eb("users.connectedAccountId", "=", ownerAccountId).and(
				"users.ownerAccountId",
				"<>",
				ownerAccountId,
			),
		)
		.innerJoin("receiptParticipants", (jb) =>
			jb.onRef("receiptParticipants.userId", "=", "users.id"),
		)
		.innerJoin("receipts", (jb) =>
			jb.onRef("receipts.id", "=", "receiptParticipants.receiptId"),
		);
