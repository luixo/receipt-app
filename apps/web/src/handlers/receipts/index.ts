import { procedure as add } from "./add";
import { procedure as get } from "./get";
import { procedure as getPaged } from "./get-paged";
import { procedure as remove } from "./remove";
import { procedure as update } from "./update";

export const router = {
	get,
	getPaged,
	remove,
	add,
	update,
};
