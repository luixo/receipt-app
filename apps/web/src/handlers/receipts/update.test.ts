import { faker } from "@faker-js/faker";
import { describe, expect } from "vitest";

import type { TRPCMutationInput } from "~app/trpc";
import { createAuthContext } from "~tests/backend/utils/context";
import {
	insertAccount,
	insertAccountWithSession,
	insertReceipt,
	insertReceiptItem,
	insertReceiptItemConsumer,
	insertReceiptParticipant,
	insertUser,
} from "~tests/backend/utils/data";
import {
	expectDatabaseDiffSnapshot,
	expectTRPCError,
	expectUnauthorizedError,
} from "~tests/backend/utils/expect";
import type { TestContext } from "~tests/backend/utils/test";
import { test } from "~tests/backend/utils/test";
import { parsers } from "~utils/date";
import { t } from "~web/handlers/trpc";
import { getRandomCurrencyCode } from "~web/handlers/utils.test";

import { procedure } from "./update";
import {
	verifyCurrencyCode,
	verifyIssued,
	verifyName,
	verifyReceiptId,
} from "./utils.test";

const createCaller = t.createCallerFactory(t.router({ procedure }));

const runTest = async (
	ctx: TestContext,
	getUpdate: () => TRPCMutationInput<"receipts.update">["update"],
) => {
	const { sessionId, accountId } = await insertAccountWithSession(ctx);
	const { id: receiptId } = await insertReceipt(ctx, accountId);
	const { id: userId } = await insertUser(ctx, accountId);
	const { id: receiptItemId } = await insertReceiptItem(ctx, receiptId);
	await insertReceiptParticipant(ctx, receiptId, userId);
	await insertReceiptItemConsumer(ctx, receiptItemId, userId);

	// Verify unrelated data doesn't affect the result
	const { id: anotherUserId } = await insertUser(ctx, accountId);
	const { id: anotherReceiptId } = await insertReceipt(ctx, accountId);
	await insertReceiptParticipant(ctx, anotherReceiptId, anotherUserId);
	await insertReceiptItem(ctx, anotherReceiptId);

	const { id: foreignAccountId } = await insertAccount(ctx);
	const { id: foreignUserId } = await insertUser(ctx, foreignAccountId);
	const { id: foreignReceiptId } = await insertReceipt(ctx, foreignAccountId);
	await insertReceiptParticipant(ctx, foreignReceiptId, foreignUserId);
	await insertReceiptItem(ctx, foreignReceiptId);

	const caller = createCaller(await createAuthContext(ctx, sessionId));
	const result = await expectDatabaseDiffSnapshot(ctx, () =>
		caller.procedure({ id: receiptId, update: getUpdate() }),
	);
	expect(result).toStrictEqual<typeof result>(undefined);
};

describe("receipts.update", () => {
	describe("input verification", () => {
		expectUnauthorizedError((context) =>
			createCaller(context).procedure({
				id: faker.string.uuid(),
				update: { type: "name", name: faker.lorem.words() },
			}),
		);

		verifyReceiptId(
			(context, receiptId) =>
				createCaller(context).procedure({
					id: receiptId,
					update: { type: "name", name: faker.lorem.words() },
				}),
			"",
		);

		verifyName(
			(context, name) =>
				createCaller(context).procedure({
					id: faker.string.uuid(),
					update: { type: "name", name },
				}),
			"update.",
		);

		verifyIssued(
			(context, issued) =>
				createCaller(context).procedure({
					id: faker.string.uuid(),
					update: { type: "issued", issued },
				}),
			"update.",
		);

		verifyCurrencyCode(
			(context, currencyCode) =>
				createCaller(context).procedure({
					id: faker.string.uuid(),
					update: { type: "currencyCode", currencyCode },
				}),
			"update.",
		);

		test("receipt not found", async ({ ctx }) => {
			const { sessionId, accountId } = await insertAccountWithSession(ctx);

			// Verifying adding other receipts doesn't affect the error
			await insertReceipt(ctx, accountId);

			const caller = createCaller(await createAuthContext(ctx, sessionId));
			const nonExistentReceiptId = faker.string.uuid();
			await expectTRPCError(
				() =>
					caller.procedure({
						id: nonExistentReceiptId,
						update: { type: "name", name: faker.lorem.words() },
					}),
				"NOT_FOUND",
				`No receipt found by id "${nonExistentReceiptId}".`,
			);
		});

		test("receipt is not owned by the account", async ({ ctx }) => {
			// Self account
			const {
				sessionId,
				accountId,
				account: { email },
			} = await insertAccountWithSession(ctx);
			// Foreign account
			const { id: otherAccountId } = await insertAccount(ctx);
			const { id: foreignReceiptId } = await insertReceipt(ctx, otherAccountId);

			// Verifying adding other receipts doesn't affect the error
			await insertReceipt(ctx, accountId);

			const caller = createCaller(await createAuthContext(ctx, sessionId));
			await expectTRPCError(
				() =>
					caller.procedure({
						id: foreignReceiptId,
						update: { type: "name", name: faker.lorem.words() },
					}),
				"FORBIDDEN",
				`Receipt "${foreignReceiptId}" is not owned by "${email}".`,
			);
		});
	});

	describe("functionality", () => {
		test("update name", async ({ ctx }) => {
			await runTest(ctx, () => ({ type: "name", name: faker.lorem.words() }));
		});

		test("update issued", async ({ ctx }) => {
			await runTest(ctx, () => ({
				type: "issued",
				issued: parsers.plainDate("2020-06-01"),
			}));
		});

		test("update currency code", async ({ ctx }) => {
			await runTest(ctx, () => ({
				type: "currencyCode",
				currencyCode: getRandomCurrencyCode(),
			}));
		});
	});
});
