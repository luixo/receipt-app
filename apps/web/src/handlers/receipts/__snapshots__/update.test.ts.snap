// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`receipts.update > functionality > update currency code 1`] = `
{
  "updated": {
    "receipts": {
      "id:f1becfdb-af87-4219-a543-ad411d7da19a": {
        "currencyCode": "NOK",
        "updatedAt": "2020-01-01 00:01:00+00",
      },
    },
  },
}
`;

exports[`receipts.update > functionality > update issued 1`] = `
{
  "updated": {
    "receipts": {
      "id:a9cf95b8-ac8a-4624-912b-de8fc0eb9637": {
        "issued": "2020-06-01",
        "updatedAt": "2020-01-01 00:01:00+00",
      },
    },
  },
}
`;

exports[`receipts.update > functionality > update name 1`] = `
{
  "updated": {
    "receipts": {
      "id:1bbf158c-dec0-45e5-aeb8-29abfd829956": {
        "name": "vitae commodi dolor",
        "updatedAt": "2020-01-01 00:01:00+00",
      },
    },
  },
}
`;
